crates/cacache/src/content/write.rs

```diff
@@ -1,54 +1,106 @@
 use std::fs::DirBuilder;
-use std::io::prelude::*;
-use std::path::{Path, PathBuf};
-use std::pin::Pin;
-use std::sync::Mutex;
-
-use async_std::fs as afs;
-use async_std::future::Future;
-use async_std::task::{self, Context, JoinHandle, Poll};
-use futures::io::AsyncWrite;
-use futures::prelude::*;
+use std::io::Seek;
+use std::io::Write;
+use std::path::PathBuf;
+
 use memmap::MmapMut;
 use ssri::{Algorithm, Integrity, IntegrityOpts};
 use tempfile::NamedTempFile;
 
 use crate::content::path;
 use crate::errors::{Internal, Result};
 
-pub const MAX_MMAP_SIZE: usize = 1024 * 1024;
+pub const MAX_MMAP_WRITE_SIZE: usize = 1024 * 1024 * 10;
+pub const MIN_MMAP_WRITE_SIZE: usize = 1024 * 1024;
 
 pub struct Writer {
     cache: PathBuf,
     builder: IntegrityOpts,
-    mmap: Option<MmapMut>,
+    target: snap::write::FrameEncoder<MaybeMmap>,
+    expected_size: Option<usize>,
+    written: usize,
+}
+
+#[derive(Debug)]
+struct MaybeMmap {
+    mmap: Option<(MmapMut, usize)>,
     tmpfile: NamedTempFile,
 }
 
+impl Seek for MaybeMmap {
+    fn seek(&mut self, pos: std::io::SeekFrom) -> std::io::Result<u64> {
+        if let Some((_, position)) = self.mmap.as_mut() {
+            match pos {
+                std::io::SeekFrom::Start(xs) => {
+                    *position = xs as usize;
+                    Ok(xs)
+                }
+                _ => unimplemented!(),
+            }
+        } else {
+            self.tmpfile.seek(pos)
+        }
+    }
+}
```

Comment:
We're using this internally so we can return to the start of the object and add the uncompressed file size. We never intend to seek from any position but `SeekFrom::Start`.